casper/src/test/scala/coop/rchain/casper/ValidateTest.scala

```diff
@@ -391,6 +391,33 @@ class ValidateTest
       } yield result
   }
 
+  "Repeat deploy validation" should "return valid for empty blocks" in withStorage {
+    implicit blockStore => implicit blockDagStorage =>
+      for {
+        _      <- createChain[Task](2)
+        block  <- blockDagStorage.lookupByIdUnsafe(0)
+        block2 <- blockDagStorage.lookupByIdUnsafe(1)
+        dag    <- blockDagStorage.getRepresentation
+        _      <- Validate.repeatDeploy[Task](block, dag, 50) shouldBeF Right(Valid)
+        _      <- Validate.repeatDeploy[Task](block2, dag, 50) shouldBeF Right(Valid)
+      } yield ()
+  }
+
+  it should "not accept blocks with a repeated deploy" in withStorage {
+    implicit blockStore => implicit blockDagStorage =>
+      for {
+        deploy  <- ConstructDeploy.basicProcessedDeploy[Task](0)
+        genesis <- createGenesis[Task](deploys = Seq(deploy))
+        block1 <- createBlock[Task](
+                   Seq(genesis.blockHash),
+                   genesis,
```

Comment:
Why does the underlying BlockDagStore require a `genessis` parameter for insert?

```scala


trait BlockDagStorage[F[_]] {
  def getRepresentation: F[BlockDagRepresentation[F]]
  def insert(
      block: BlockMessage,
      genesis: BlockMessage,
      invalid: Boolean
  ): F[BlockDagRepresentation[F]]
```

Reply:
Whenever we bond a new validator, their latest message needs to be set to the genesis block and so we pass it in as the BlockDagRepresentation itself isn't aware of what is the genesis block.